RFC for Kernel Templates (#2301)
uxlfoundation/oneDPL: rfcs/experimental/kernel_templates/README.md

# Kernel Templates

## Introduction

This RFC describes an experimental set of algorithms with easily tunable controls,
designed to achieve optimal performance on specific hardware and workloads.
They prioritize efficiency over generality when compared to
algorithms that conform to standard C++ interfaces and use device execution policies.

These algorithms are intended for use with SYCL.

The terms "kernel template" and "algorithm" are used interchangeably in this RFC.

## Status

> Kernel Templates proposal is expected to evolve into an extension.

Reviewer: An extension of what?

I guess you mean that there is no intention to add KT into the oneDPL specification. But that might change, and I am not sure if it is important in the context of this RFC.

Author: I've removed that note.

      /*return*/ reduce(/*args*/);
      /*return*/ transform(/*args*/);
      // ...
    }
  } // namespace oneapi::dpl::experimental::kt::gpu

  namespace cpu
  {
    // Algorithms optimized for CPU architectures
    // ...
  }

} // namespace oneapi::dpl::experimental::kt
```

There is no requirement for the sets of algorithms in each namespace to be identical.

### Abstract Algorithm Signature

A kernel template is a C++ function invoked from the host.

Each function takes a `sycl::queue` as its first argument and
an instance of the [kernel_param](kernel_configuration/README.md) template,
which applies settings common to all kernel templates, as its last argument.
The arguments in between vary depending on the specific algorithm.
The function may also have algorithm-specific non-deducible template parameters.

Below is an abstract signature of a kernel template:

```c++
// defined in <oneapi/dpl/experimental/kernel_templates>
// in namespace oneapi::dpl::experimental::kt

template <
    typename AlgorithmParameter1,  // manually specified, may have a default
    typename AlgorithmParameter2,  // ...
    typename AlgorithmParameterN,  // manually specified, may have a default

>     typename KernelParam,          // manually specified, may have a default in future

Author: Corrected

including the number of kernels launched, register and local memory usage,
global memory access, and the utilization of hardware computational resources.
They can be easily adjusted for another GPU with different hardware characteristics.

The kernel template relies on ESIMD technology
and certain forward progress guarantees between work-groups.

```c++
// icpx -fsycl radix_sort.cpp -o radix_sort -I /path/to/oneDPL/include && ./radix_sort
#include <cstdint>
#include <iostream>
#include <sycl/sycl.hpp>

#include <oneapi/dpl/experimental/kernel_templates>

namespace kt = oneapi::dpl::experimental::kt;

int main()
{
   std::size_t n = 6;
   sycl::queue q{sycl::gpu_selector_v};
   std::uint32_t* keys = sycl::malloc_shared<std::uint32_t>(n, q);

   keys[0] = 3, keys[1] = 2, keys[2] = 1, keys[3] = 5, keys[4] = 3, keys[5] = 3;

   auto e = kt::gpu::esimd::radix_sort<false, 8>(q, keys, keys + n, kt::kernel_param<416, 64>{});
   e.wait();

   sycl::free(keys, q);
   return 0;
}
```

## Testing

The testing harness should cover:
 - Varying numbers of elements to process, from thousands to millions.
 - Different element types.
 - Diverse types of input sequences.

>  - Representative data distributions.

Author: I've added examples of what I meant under `Different element types` and `Diverse types of input sequences` points and removed `Representative data distributions`.

The later is too obvious, for example: the data should be unsorted before calling the sort algorithm being tested, or if an algorithm is searching something, there should be data to find in one case or nothing in another case and etc.

>  - Representative data distributions.

Reviewer: Looks good now, thanks.

 - Various combinations of other parameters, depending on the signature of the algorithm.
 - Edge cases (for example, zero-sized sequences).

Due to the unlimited number of possible kernel parameter combinations, it is recommended to:
 - Always test the most representative configurations
  (most likely to be used and achieve the best performance).
 - Randomly select a subset of additional parameter combinations to broaden coverage.

The testing harness should generate or enable tests with limited portability on demand,
for example through cmake arguments.

## Open Questions

### Name

The name "Kernel Templates" may be misleading because
these entities more act like algorithms rather than SYCL kernels.
Probably, a better name is "Algorithm Templates".
The renaming requires changing the corresponding namespace.

### Specializations and their Differentiation

Currently, the specializations of the algorithms belong to different namespaces.
Using tags instead of namespaces offers several advantages,

> for example easier dispatching between specializations and avoiding deeply nested namespaces.

Author: Yes, it was unintentional, and thank you giving such feedback - this is what I was looking for. I've updated that section. Now, the approach with namespaces looks much more preferrable, but we can keep that section to make a final decision after adding more algorithms.

by providing an easy way to get memory requirements and apply fallbacks if needed.

### External Allocation of Global Memory

There are scenarios which benefit from passing pre-allocated temporary memory to an algorithm.
For example, a user program may manage a reusable memory pool.
Also, this is especially useful when an algorithm is called in a loop and the allocation overhead
cannot be amortized by memory pools provided by the device or compiler runtime.

It may be preferable to make this the only option and
avoid internal global memory allocations altogether.
This approach eliminates the overhead associated with `host_task`,
which retains temporary memory until an algorithm completes its asynchronous execution.
Additionally, it reduces the number of interfaces that need to be supported.

This should be done in conjunction with the
[Global and Local Memory Requirements](#reporting-global-and-local-memory-requirements) interface.

### Asynchronous Execution

The algorithms return a `sycl::event` but do not accept any input events.
As a result, they are not fully asynchronous.

It should be investigated whether full asynchrony is desirable,

> and if so, how to implement it correctly.

Reviewer: What do you mean by "full asynchrony", and why is it only "partial" without input events?

It might be more appropriate to speak about supporting explicit dependency management.

Author: Corrected.

I meant that the algorithms cannot be used in asynchronous chains as a non-last point. That limitation does not make them "partially" asynchronous, so thanks for the comment.

### Separation of Specializations Based on a Problem Size

Different approaches may be more effective when processing different number of elements.
For example, a small number of elements can be processed by a single work-group.

Separating such cases would reduce compilation time.

### Algorithms to Implement

A list of algorithms to implement should be defined.

> 

Reviewer: Why is this an open question? What purpose would the list of algorithms serve?

Author: It would serve as a plan which algorithms to implement. It can be useful, e.g. for external contributors as a suggestion what can definitely be helpful for oneDPL.

However, the structure being proposed can have many specializations which cannot be predicted (e.g. future architectures), it is difficult to define the list. Perhaps, it can be a minimal list of algorithms in the portable namespace.

Author: Also, it can be one of the indicators of maturity of the Kernel Templates for it to exit the experimental status.

Author: I definitely should make the description more precise or remove the section entirely, depending on what we decide here. I'd like to get your opinion first, e.g. if my answer regarding the purpose of the list is compelling.

Author: Could you check  "Algorithms to Implement" section added in https://github.com/uxlfoundation/oneDPL/pull/2301/commits/f93e4a604d76c586f850006922f8618158164922?

However, it is unclear if a default value should be provided,
especially given the extension's focus on performance tuning.

Creation of a default value with optimal settings is impossible right now.
`kernel_param` template argument, which substitutes `KernelParam`,
includes compile-time values, whose optimal selection depends on run-time information,
such is the number of elements to process.

[Runtime Parameters](kernel_configuration/README.md#runtime-parameters) raises a question
of providing them at run-time.
However, doing so would likely require querying device properties, introducing overhead.

The complexity of manual selection of these values may be reduced by the [Benchmarks](#benchmarking),
making default values less relevant.

### Configuring an Algorithm with Multiple Kernels

It needs to be explored how to configure algorithms with multiple performance-critical kernels.
The `kernel_param` targets a single kernel, and currently, only one instance can be passed.

### Compiler Extension and Differentiation of Algorithms

Currently, it is unclear if we should defferentiate algorithms
which rely compiler extensions.

## Exit Criteria

> The proposed set of algorithms should transition to an extension if:

Reviewer: Here, again, let's speak about transition into production, leaving aside whether it's "an extension" or something else.

> - All the questions under [Kernel Templates](#open-questions) and
>   [Kernel Configuration](kernel_configuration/README.md#open-questions) sections are addressed,
>   unless a question is marked as not mandatory.

Reviewer: ```suggestion
- All the questions under [Kernel Templates](#open-questions) and
  [Kernel Configuration](kernel_configuration/README.md#open-questions) sections are either addressed
  or provided with a justification to postpone or ignore.
```

- A significant portion of the algorithms listed in
  [Algorithms to Implement](#algorithms-to-implement) is implemented.

> - They are performant.

Reviewer: ```suggestion
- Evidence of sufficiently good performance is provided.
```

Below is an abstract signature of a kernel template:

```c++
// defined in <oneapi/dpl/experimental/kernel_templates>
// in namespace oneapi::dpl::experimental::kt

template <
    typename AlgorithmParameter1,  // manually specified, may have a default
    typename AlgorithmParameter2,  // ...
    typename AlgorithmParameterN,  // manually specified, may have a default

    typename KernelParam,          // deduced, may have a default in future

    typename Arg1,                 // deduced, may have a default
    typename Arg2,                 // ...
    typename ArgN                  // deduced, may have a default
>
sycl::event kernel_template (
    sycl::queue q,
    /* any cvref */ Arg1 arg1,
    /* any cvref */ Arg2 arg2,
    /* any cvref */ ArgN argn,
    KernelParam param              // manually specified, may have a default in future
);
```

`Arg1`, ..., `ArgN` include sequences to process.
Data must be passed using the same mechanisms described in the
[documentation on passing data](https://uxlfoundation.github.io/oneDPL/parallel_api/pass_data_algorithms.html#pass-data-to-algorithms)
for algorithms with device policies.

> Specialized algorithms may extend or restrict the supported data passing mechanisms.

Reviewer 2: Do we intend to support host allocated data in Kernel templates?
  
I'd suggest we require data to be indirectly device accessible, given that these are intended to favor performance over productivity and this is sycl specific.  However, if we want to support buffers with `begin/end` (which I think we do), we will still need to process the input, so it may not make much of a technical difference in ease of implementation.  We may want to detect and exclude "bad" data, and / or process it a little differently from mainline oneDPL algorithms.

and describes how it is tuned for better performance and what makes it less general
than an alternative algorithm with a standard interface.
It uses `oneapi::dpl::experimental::kt::gpu::esimd::radix_sort`,
which can be compared to `oneapi::dpl::stable_sort`.

Performance tuning controls:
1. Algorithm-specific: the number of bits sorted per radix sort pass (`8`).
2. Common: the number of elements processed per work-item (`416`)
  and the number of work-items in a work-group (`64`).

These parameters influence various factors,
including the number of kernels launched, register and local memory usage,
global memory access, and the utilization of hardware computational resources.
They can be easily adjusted for another GPU with different hardware characteristics.

The kernel template relies on ESIMD technology
and certain forward progress guarantees between work-groups.

```c++
// icpx -fsycl radix_sort.cpp -o radix_sort -I /path/to/oneDPL/include && ./radix_sort
#include <cstdint>
#include <iostream>
#include <sycl/sycl.hpp>

#include <oneapi/dpl/experimental/kernel_templates>

namespace kt = oneapi::dpl::experimental::kt;

int main()
{

>    std::size_t n = 6;

Reviewer 3: ```suggestion
   const std::size_t n = 6;
```

a potential addition of a default value in the future.
However, it is unclear if a default value should be provided,
especially given the focus on performance tuning.

Creation of a default value with optimal settings is impossible right now.
`kernel_param` template argument, which substitutes `KernelParam`,
includes compile-time values, whose optimal selection depends on run-time information,
such is the number of elements to process.

[Runtime Parameters](kernel_configuration/README.md#runtime-parameters) raises a question
of providing them at run-time.
However, doing so would likely require querying device properties, introducing overhead.

The complexity of manual selection of these values may be reduced by the [Benchmarks](#benchmarking),
making default values less relevant.

### Configuring an Algorithm with Multiple Kernels

It needs to be explored how to configure algorithms with multiple performance-critical kernels.
The `kernel_param` targets a single kernel, and currently, only one instance can be passed.

### Compiler Extension and Differentiation of Algorithms

Currently, it is unclear if we should defferentiate algorithms
which rely compiler extensions.

> 

Author: I'll add a section for KT as a backend for the oneDPL mainline. I like the idea with the macro as both a mechanism to express the requirement on a code level and as a knob for oneDPL mainline.

  (most likely to be used and achieve the best performance).
 - Randomly select a subset of additional parameter combinations to broaden coverage.

The testing harness should generate or enable tests with limited portability on demand,
for example through cmake arguments.

## Open Questions

### Name

The name "Kernel Templates" may be misleading because
these entities more act like algorithms rather than SYCL kernels.
Renaming would require changing the top-level namespace.

### Specializations and their Differentiation

Currently, the specializations of the algorithms belong to different namespaces.

Using a tag as a function or a template parameter should be considered as well.
This method has both advantages and disadvantages compared to namespace-based separation.
Pros:
- Avoids potential name clashes caused by `using` directives.
Cons:
- Does not replicate the hierarchical structure provided by namespaces,
  unless namespeces are built into the tag name.
Other aspects:
- Is an additional parameter, possibly with its own namespaces,
  preferred over adding namespaces directly to an algorithm?
- Specializations may not share the same set of arguments,

>   which reduces the value of tag-based dispatching.

Reviewer: The lists need to be separated with empty lines for proper rendering by GitHub.

```

`Arg1`, ..., `ArgN` include sequences to process.
These arguments must reference data which is ready for processing;
any necessary data transfers must be completed in advance or managed automatically.
Supported data storage types include:
- [SYCL buffers](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:buffers).
- Unified Shared Memory ([USM](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:usm)).
- `std::vector` with `sycl::usm_allocator`.

> Supported data passing mechanisms are listed in
> the [documentation on passing data](https://uxlfoundation.github.io/oneDPL/parallel_api/pass_data_algorithms.html#pass-data-to-algorithms).
> The `oneapi::dpl::begin` and `oneapi::dpl::end` helper functions are not supported;
> however, `sycl::buffer` objects may be passed directly.

Reviewer 2: To make it clear I think we need one more disclaimer to the info in the link, assuming I understand the intent appropriately:

```suggestion
Supported data passing mechanisms are listed in
the [documentation on passing data](https://uxlfoundation.github.io/oneDPL/parallel_api/pass_data_algorithms.html#pass-data-to-algorithms).
The `oneapi::dpl::begin` and `oneapi::dpl::end` helper functions are not supported;
however, `sycl::buffer` objects may be passed directly. `std::vector` with host-allocated data are not supported.
```

Author: You are right, thanks for finding that. I am going to apply a slightly corrected suggestion:

```suggestion
Supported data passing mechanisms are listed in
the [documentation on passing data](https://uxlfoundation.github.io/oneDPL/parallel_api/pass_data_algorithms.html#pass-data-to-algorithms).
The `oneapi::dpl::begin` and `oneapi::dpl::end` helper functions are not supported;
however, `sycl::buffer` objects may be passed directly.
`std::vector` with host-allocated data is not supported.
```

Currently, it is unclear if we should defferentiate algorithms
which rely on compiler extensions.

### Kernel Templates as a Backend for Algorithms with Standard Interfaces

It should be evaluated whether the proposed algorithms can serve as a backend for oneDPL algorithms
that use the standard C++ interfaces and device execution policies.

It will affect which data-passing mechanisms,
as described in [Abstract Algorithm Signature](abstract-algorithm-signature) section, are allowed.
In particular, it should be investigated if supporting these entities is necessary and
whether they can be handled by an additional thin layer between
the algorithms with standard interfaces and Kernel Templates:
- `oneapi::dpl::begin` and `oneapi::dpl::end`, which mimic iterators to pass `sycl::buffer` to algorithms accepting iterators.
- Iterators to a host-allocated `std::vector` as described in

>   [Pass Data to Algorithms > Use std::vector](https://github.com/uxlfoundation/oneDPL/blob/release/2022.9.0/documentation/library_guide/parallel_api/pass_data_algorithms.rst#use-stdvector)

Reviewer: At glance, it does not seem necessary for KT, even if it serves as the implementation backend for the "regular" algorithms. The necessary data copying/transferring to the device can and likely should be performed before KT is used.

Author: I've removed this paragraph in the KT as a backed sub-section, and updated the main description on data passing mechanisms. See https://github.com/uxlfoundation/oneDPL/pull/2301/commits/5001928fdddcaf30cff26418e1bbb6fb59e0336f.

In particular, it should be investigated if supporting these entities is necessary and
whether they can be handled by an additional thin layer between
the algorithms with standard interfaces and Kernel Templates:
- `oneapi::dpl::begin` and `oneapi::dpl::end`,
   which mimic iterators to pass `sycl::buffer` to iterator-based algorithms.
- Iterators to a host-allocated `std::vector` as described in
  [Pass Data to Algorithms > Use std::vector](https://github.com/uxlfoundation/oneDPL/blob/release/2022.9.0/documentation/library_guide/parallel_api/pass_data_algorithms.rst#use-stdvector)

An additional question is how to handle hardware requirements that cannot be queried directly
through standard SYCL mechanisms. For example, selecting a more performant kernel template
might depend on forward-progress guarantees between work-groups.
While oneAPI DPC++ supports querying this via
the [Root Group](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_root_group.asciidoc)

> experimental extension, other SYCL implementations may lack a corresponding feature.

Reviewer: That's not a problem I think - at least not for the use of KT as a backend. Implementations with HW requirements like forward progress are supposed to have a fallback path - either another KT algorithm or a "standard SYCL" implementation. If we cannot prove that the requirements are met for any reason, whether HW or SYCL implementation related, the fallback should be used.

Furthermore, even if a device does not explicitly support cooperative kernel launch,
it may still provide sufficient guarantees in limited cases
such as chain-like synchronization between work-groups.
A possible solution is to allow users to define a macro

> indicating that the required hardware capability is present.

Reviewer: One of the motivations for KT is to give users a choice if the existing API does not meet their needs. If their HW provides "sufficient" guarantees, KT can be used directly.

> ### Encoding Requirements

Reviewer: Honestly, I do not quite understand this whole subsection. It seems as yet another way to mention (and generalize?) the progress guarantee problem with yet another macro-based solution proposed. As written, I find it too abstract to be useful.

Please try to improve it, so that the specifics of the concern and/or the idea is clear, and is different from what has already been discussed. And ideally, avoid this "let's add an opt-in macro" approach to solve - it has too many drawbacks, the main ones being lack of granularity and the risk of multiple function definition at linkage (aka ODR violation).

Reviewer 2: This came from my suggestion.  Perhaps it is not worth the downsides.  But I will list a couple of advantages:
1) It can allow users to be less conservative if they so choose (For instance, hardware which is does not explicitly have forward progress guarantees but does in practice work without hangs)
2) It can avoid having to do unnecessary JIT compilation for fallbacks which wont be used
3) It allows users to not need to rewrite their code to use KT to take advantage of them

However, with the issues you bring up, perhaps these are not worthwhile, and users who want 1 + 2 can just use the explicit KT API.

Reviewer: Looking closer at these sentences:
> Some requirements, for example forward-progress guarantees between work-groups, are currently only described in documentation. Encoding such requirements programmatically would improve visibility and help avoid unintended usage. 

I feel like maybe @dmitriy-sobolev tries to ask the question: For things like forward progress guarantees and ESIMD support expected by a certain function, can we enforce the HW/SW requirements better than just stating those in the docs?

We can discuss several possible sides of that. Seemingly the further text in the subsection, as @danhoeflinger mentions, is about some opt-in mechanisms to allow using these "demanding" implementations in the backend if those would not be chosen otherwise. Though that's not "avoiding unintended usage" but rather enforcing intended usage. I tend to think it's not worth it, both because of the downsides and because of likely niche applicability. Anyway, it is better covered in the "KT as a backend" subsection.

Another possible question is "Can the special requirements be enforced programmatically when KT algorithms are used directly?" I think it depends on how generic and portable we want these algorithms to be. And the current mechanism to control that is the namespace hierarchy. For example, the radix sort implementation for platforms that support ESIMD is placed in the `kt::gpu::esimd` namespace. while a more portable implementation can be in the `kt::gpu` namespace. 

We can change the API, e.g., use tags or additional options in `kernel_config`, or modify algorithm names (that could be the preference for single work group implementation for example); but in essence it's all the same - there is already compile-time opt-in for users to indicate exactly what they want if choice is available. I do not think we can do better requirement enforcement / prevention of unintended usage in the compile time.
